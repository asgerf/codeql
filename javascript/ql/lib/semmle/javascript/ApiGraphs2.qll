import javascript
private import semmle.javascript.dataflow.internal.FlowSteps as FlowSteps
private import internal.CachedStages
private import semmle.javascript.DeepFlow
private import semmle.javascript.dataflow.internal.DataFlowNode

module API2 {
  deprecated Node root() { none() }

  /**
   * A class for contributing new steps for tracking uses of an API.
   */
  class AdditionalUseStep extends Unit {
    /**
     * Holds if use nodes should flow from `pred` to `succ`.
     */
    predicate step(DataFlow::SourceNode pred, DataFlow::SourceNode succ) { none() }
  }

  private module AdditionalUseStep {
    pragma[nomagic]
    predicate step(DataFlow::SourceNode pred, DataFlow::SourceNode succ) {
      any(AdditionalUseStep st).step(pred, succ)
    }
  }

  cached
  Node moduleImport(string name) {
    result = DataFlow::moduleImport(name)
    or
    result = DataFlow::moduleMember(name, "default")
  }

  module Node {
    Node ofType(string packageName, string typeName) {
      result.(DataFlow::SourceNode).hasUnderlyingType(packageName, typeName)
    }
  }

  class InvokeNode extends DataFlow::InvokeNode {
    pragma[inline]
    SinkNode getParameter(int i) { result = this.getArgument(i) }

    pragma[inline]
    SinkNode getAParameter() { result = this.getAnArgument() }

    pragma[inline]
    SinkNode getLastParameter() { result = this.getLastArgument() }

    pragma[inline]
    Node getReturn() { result = this }

    pragma[inline]
    Node getInstance() { result = this and this instanceof DataFlow::NewNode }
  }

  class CallNode extends InvokeNode, DataFlow::CallNode { }

  class NewNode extends InvokeNode, DataFlow::NewNode { }

  class Node extends TNode instanceof DataFlow::SourceNode {
    deprecated Node getASuccessor() { none() }

    string toString() { result = super.toString() }

    predicate hasLocationInfo(
      string filepath, int startline, int startcolumn, int endline, int endcolumn
    ) {
      super.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
    }

    pragma[inline]
    private DataFlow::SourceNode forward() {
      Deep::hasFlowTo(pragma[only_bind_out](this), pragma[only_bind_into](result))
    }

    pragma[inline]
    DataFlow::Node getAValueReachableFromSource() { result = this.forward().getALocalUse() }

    pragma[inline]
    DataFlow::SourceNode asSource() { result = this }

    pragma[inline]
    API2::CallNode getACall() { result = this.forward().getACall() }

    pragma[inline]
    API2::CallNode getMaybePromisifiedCall() {
      result = this.getACall()
      or
      result = Deep::getABoundInvocation(this, true, _)
    }

    pragma[inline]
    API2::NewNode getAnInstantiation() { result = this.forward().getAnInstantiation() }

    pragma[inline]
    API2::InvokeNode getAnInvocation() { result = this.forward().getAnInvocation() }

    pragma[inline]
    API2::Node getMember(string name) { result = Deep::getLoad(pragma[only_bind_out](this), name) }

    pragma[inline]
    API2::Node getUnknownMember() {
      DataFlow::SourceNode::Internal::dynamicPropRef(pragma[only_bind_out](this).forward(),
        pragma[only_bind_into](result)) and
      pragma[only_bind_out](result) instanceof DataFlow::PropRead
      // Note: The type check above is not strictly needed since API::Node is a subtype of SourceNode, and PropWrite is not a SourceNode.
      // TODO: Investigate which check is more efficient in practice.
    }

    pragma[inline]
    API2::Node getAMember() { result = pragma[only_bind_out](this.forward()).getAPropertyRead() }

    pragma[inline]
    API2::Node getInstance() {
      result = this.getAnInstantiation() // note: these predicates return the same thing, but with different types
    }

    pragma[inline]
    API2::SinkNode getParameter(int i) {
      Deep::argumentPassing(pragma[only_bind_out](this), i, result)
    }

    pragma[inline]
    API2::SinkNode getLastParameter() {
      result = pragma[only_bind_out](this.getACall()).getLastArgument()
      or
      result =
        TApiSyntheticCallbackArg(Deep::getABoundInvocation(pragma[only_bind_out](this), true, _))
    }

    pragma[inline]
    API2::SinkNode getAParameter() { Deep::argumentPassing(pragma[only_bind_out](this), _, result) }

    pragma[inline]
    API2::SinkNode getReceiver() {
      Deep::receiverPassing(this, result) // TODO: change to only return explicit receivers?
    }

    pragma[inline]
    API2::Node getReturn() {
      result = this.getAnInvocation() // note: these predicates return the same thing, but with different types
    }

    pragma[inline]
    API2::Node getPromised() { result = Deep::getPromised(this) }

    pragma[inline]
    API2::Node getPromisedError() { result = Deep::getPromisedError(this) }

    Node getADecoratedClass() { none() } // TODO

    Node getADecoratedMember() { none() } // TODO

    SinkNode getADecoratedParameter() { none() } // TODO

    /**
     * DEPRECATED. Use `getACall().getNumArgument()` instead, but note that each call may have a different number of arguments.
     *
     * This predicate returns the largest arity among these calls, or zero if there are no calls.
     */
    deprecated int getNumParameter() { result = max(int s | exists(this.getParameter(s))) + 1 }

    /** DEPRECATED. `API::Node` can no longer represent sinks -- use `API::SinkNode` instead. */
    deprecated DataFlow::Node asSink() { none() }

    /** DEPRECATED. `API::Node` can no longer represent sinks -- use `API::SinkNode` instead. */
    deprecated DataFlow::SourceNode getAValueReachingSink() { none() }

    /** DEPRECATED. This predicate has been renamed to `asSource`. */
    deprecated DataFlow::SourceNode getAnImmediateUse() { result = this.asSource() }

    /** DEPRECATED. This predicate has been renamed to `getAValueReachableFromSource`. */
    deprecated DataFlow::Node getAUse() { result = this.getAValueReachableFromSource() }

    /** DEPRECATED. This predicate has been renamed to `asSink` and moved to the `API::SinkNode` class. */
    deprecated DataFlow::Node getARhs() { result = this.asSink() }

    /** DEPRECATED. This predicate has been renamed to `getAValueReachingSink` and moved to the `API::SinkNode` class. */
    deprecated DataFlow::Node getAValueReachingRhs() { result = this.getAValueReachingSink() }
  }

  class BacktrackingNode extends TNode instanceof DataFlow::SourceNode {
    // TODO: include synth callback arg
    string toString() { result = this.(DataFlow::Node).toString() }

    predicate hasLocationInfo(
      string filepath, int startline, int startcolumn, int endline, int endcolumn
    ) {
      this.(DataFlow::Node).hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
    }

    pragma[inline]
    private DataFlow::SourceNode backward() {
      Deep::hasFlowTo(pragma[only_bind_into](result), pragma[only_bind_out](this))
    }

    pragma[inline]
    DataFlow::SourceNode getAValueReachingSink() { result = this.backward() }

    pragma[inline]
    DataFlow::FunctionNode getAFunction() { result = this.backward() }

    pragma[inline]
    API2::SinkNode getMember(string name) {
      // TODO: include store edges for class members
      // TODO: include reverse-load edges for stepping through long access paths
      result = this.backward().getAPropertyWrite(name).getRhs()
    }

    pragma[inline]
    API2::SinkNode getUnknownMember() {
      exists(DataFlow::PropWrite write |
        DataFlow::SourceNode::Internal::dynamicPropRef(pragma[only_bind_out](this).backward(),
          pragma[only_bind_into](write)) and
        result = write.getRhs()
      )
    }

    pragma[inline]
    API2::SinkNode getAMember() {
      // TODO: include store edges for class members
      result = this.backward().getAPropertyWrite().getRhs()
    }

    pragma[inline]
    API2::Node getInstance() { result = Deep::getClassReceiverRef(this.backward()) }

    pragma[inline]
    API2::Node getParameter(int i) { Deep::parameterDef(this.backward(), i, result) }

    pragma[inline]
    API2::Node getAParameter() { Deep::parameterDef(this.backward(), _, result) }

    pragma[inline]
    API2::Node getLastParameter() {
      // TODO: is this the best way to do this?
      exists(DataFlow::SourceNode src | src = this.backward() |
        result = src.(DataFlow::FunctionNode).getLastParameter()
        or
        result = src.(DataFlow::ClassNode).getConstructor().getLastParameter()
      )
    }

    pragma[inline]
    API2::Node getReceiver() { result = this.getAFunction().getReceiver() }

    pragma[inline]
    API2::SinkNode getReturn() { result = Deep::getPrettyReturn(this.getAFunction()) }

    pragma[inline]
    API2::SinkNode getPromised() {
      result = Deep::getStoreRhs(this.backward(), Promises::valueProp())
    }

    pragma[inline]
    API2::SinkNode getPromisedError() {
      result = Deep::getStoreRhs(this.backward(), Promises::errorProp())
    }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated API::Node getADecoratedClass() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated API::Node getADecoratedMember() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated API::Node getADecoratedParameter() { none() }

    /**
     * DEPRECATED. Use `getAFunction().getNumParameter()` instad, but note that different functions flowing here may
     * each have a different number of parameters.
     *
     * This predicate returns the largest number of parameters, or zero if there are no functions flowing here.
     */
    deprecated int getNumParameter() { result = max(int s | exists(this.getParameter(s))) + 1 }

    /**
     * DEPRECATED. This predicate should only be used on `API::SinkNode`. Once backtracking is performed the node is rarely
     * useful as a sink.
     *
     * Returns the receiver itself as a data-flow node, though this does not generally have the appropriate location for u
     */
    deprecated DataFlow::Node asSink() { result = this }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::CallNode getAnInvocation() { none() }

    /**
     * DEPRECATED. This predicate has no result for backtracking nodes.
     * - For finding references to `this` inside a class, use `getInstance()` instead.
     * - For finding instantiation sites of a class, use a forward-tracking node instead (`API::Node`).
     */
    deprecated DataFlow::NewNode getAnInstantiation() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::CallNode getACall() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::CallNode getMaybePromisifiedCall() { none() }

    /** DEPRECATED. This predicate has been renamed to `asSource` and is only usable for forward-tracking nodes (`API::Node`). */
    deprecated DataFlow::SourceNode getAnImmediateUse() { none() }

    /** DEPRECATED. This predicate has been renamed to `getAValueReachableFromSource` and is only usable for forward-tracking nodes (`API::Node`). */
    deprecated DataFlow::Node getAUse() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::SourceNode asSource() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::SourceNode getAValueReachableFromSource() { none() }

    /** DEPRECATED. This predicate has been renamed to `asSink` and should only be use on the `API::SinkNode` class. */
    deprecated DataFlow::Node getARhs() { result = this.asSink() }

    /** DEPRECATED. This predicate has been renamed to `getAValueReachingSink`. */
    deprecated DataFlow::Node getAValueReachingRhs() { result = this.getAValueReachingSink() }
  }

  class SinkNode extends TDataFlowNodeOrApiNode {
    // TODO: include synth callback arg
    string toString() { result = this.(DataFlow::Node).toString() }

    predicate hasLocationInfo(
      string filepath, int startline, int startcolumn, int endline, int endcolumn
    ) {
      this.(DataFlow::Node).hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
    }

    pragma[inline]
    API2::BacktrackingNode backtrack() {
      result = pragma[only_bind_out](this).(DataFlow::Node).getALocalSource()
    }

    pragma[inline]
    DataFlow::SourceNode getAValueReachingSink() {
      result = this.backtrack().getAValueReachingSink()
    }

    pragma[inline]
    API2::SinkNode getMember(string name) { result = this.backtrack().getMember(name) }

    pragma[inline]
    API2::SinkNode getUnknownMember() { result = this.backtrack().getUnknownMember() }

    pragma[inline]
    API2::SinkNode getAMember() { result = this.backtrack().getAMember() }

    pragma[inline]
    API2::Node getInstance() { result = this.backtrack().getInstance() }

    pragma[inline]
    API2::Node getParameter(int i) {
      result = this.backtrack().getParameter(i)
      or
      Deep::promisifiedCallbackParameterDef(this, i, result)
    }

    pragma[inline]
    API2::Node getAParameter() { result = this.backtrack().getAParameter() }

    pragma[inline]
    API2::Node getLastParameter() { result = this.backtrack().getLastParameter() }

    pragma[inline]
    API2::Node getReceiver() { result = this.backtrack().getReceiver() }

    pragma[inline]
    API2::SinkNode getReturn() { result = this.backtrack().getReturn() }

    pragma[inline]
    API2::SinkNode getPromised() { result = this.backtrack().getPromised() }

    pragma[inline]
    API2::SinkNode getPromisedError() { result = this.backtrack().getPromisedError() }

    /**
     * Returns this node as a data-flow node.
     */
    pragma[inline]
    DataFlow::Node asSink() { result = this }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated API::Node getADecoratedClass() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated API::Node getADecoratedMember() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated API::Node getADecoratedParameter() { none() }

    /**
     * DEPRECATED. Use `getAFunction().getNumParameter()` instad, but note that different functions flowing here may
     * each have a different number of parameters.
     *
     * This predicate returns the largest number of parameters, or zero if there are no functions flowing here.
     */
    deprecated int getNumParameter() { result = max(int s | exists(this.getParameter(s))) + 1 }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::CallNode getAnInvocation() { none() }

    /**
     * DEPRECATED. This predicate has no result for backtracking nodes.
     * - For finding references to `this` inside a class, use `getInstance()` instead.
     * - For finding instantiation sites of a class, use a forward-tracking node instead (`API::Node`).
     */
    deprecated DataFlow::NewNode getAnInstantiation() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::CallNode getACall() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::CallNode getMaybePromisifiedCall() { none() }

    /** DEPRECATED. This predicate has been renamed to `asSource` and is only usable for forward-tracking nodes (`API::Node`). */
    deprecated DataFlow::SourceNode getAnImmediateUse() { none() }

    /** DEPRECATED. This predicate has been renamed to `getAValueReachableFromSource` and is only usable for forward-tracking nodes (`API::Node`). */
    deprecated DataFlow::Node getAUse() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::SourceNode asSource() { none() }

    /** DEPRECATED. This predicate has no result for backtracking nodes, use a forward-tracking node instead (`API::Node`). */
    deprecated DataFlow::SourceNode getAValueReachableFromSource() { none() }

    /** DEPRECATED. This predicate has been renamed to `asSink` and should only be use on the `API::SinkNode` class. */
    deprecated DataFlow::Node getARhs() { result = this.asSink() }

    /** DEPRECATED. This predicate has been renamed to `getAValueReachingSink`. */
    deprecated DataFlow::Node getAValueReachingRhs() { result = this.getAValueReachingSink() }
  }
}
