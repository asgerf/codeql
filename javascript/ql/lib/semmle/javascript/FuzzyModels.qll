private import javascript

private DataFlow::SourceNode blockedNode() {
  exists(string name | name = result.(DataFlow::PropRead).getPropertyName() |
    name =
      any(ExternalInstanceMemberDecl decl | decl.getBaseName() = ["Object", "String", "Array"])
          .getName()
    or
    name = ["then", "catch", "finally"]
  )
  or
  exists(API::AdditionalUseStep step | step.step(_, result))
  or
  DataFlow::SharedFlowStep::step(_, result)
}

/**
 * Gets a normalized package name from `importString`.
 */
bindingset[importString]
string normalizePackageName(string importString) {
  result =
    importString
        .regexpReplaceAll("^node:", "")
        .regexpReplaceAll("\\.js$", "")
        .regexpReplaceAll("/index$", "")
}

/**
 * Gets a node believed to originate from `package`.
 *
 * The package is normalized by `normalizePackageName`.
 */
API::Node getANodeFromPackage(string package) {
  exists(string importString |
    result = API::moduleImport(importString) and
    package = normalizePackageName(importString)
  )
  or
  result = getANodeFromPackage(package).getAMember() and
  not result.asSource() = blockedNode()
  or
  result = getANodeFromPackage(package).getReturn()
  or
  result = getANodeFromPackage(package).getInstance()
  or
  result = getANodeFromPackage(package).getAParameter().getAParameter()
  or
  result = getANodeFromPackage(package).getAParameter().getAMember().getAParameter()
  or
  result = getANodeFromPackage(package).getPromised()
}

/**
 * Gets a call targeting targeting a method named `methodName` in `package`.
 */
DataFlow::InvokeNode getANamedExternalCall(string package, string methodName) {
  exists(API::Node member |
    member = getANodeFromPackage(package).getMember(methodName) and
    result = member.getAnInvocation() and
    not member.asSource() = blockedNode() and
    not (methodName = "default" and member.asSource() instanceof DataFlow::ModuleImportNode)
  )
}

/**
 * Gets a call that invokes `package` as a function (that is, its exported object is a function).
 */
DataFlow::InvokeNode getADirectExternalCall(string package) {
  result = API::moduleImport(package).getAnInvocation()
}

/**
 * Gets a call target likely targets `package`, but with no associated method name.
 *
 * For example:
 * ```javascript
 * const lib = require('example')
 *
 * let fn = lib();
 * fn() // curried calls are considered anonymous
 *
 * lib(fn => {
 *   fn(); // callback invocation is considered anonymous
 * })
 *
 * lib[complicated()](); // method call with unknown name
 * ```
 */
DataFlow::InvokeNode getAnAnonymousExternalCall(string package) {
  result = getANodeFromPackage(package).getAnInvocation() and
  not result = getANodeFromPackage(package).getMember(_).getAnInvocation() and
  not result = getADirectExternalCall(package)
}
