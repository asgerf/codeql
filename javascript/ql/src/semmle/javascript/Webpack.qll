/**
 * Models the behavior of `webpack` configuration files.
 */

private import javascript

private class WebpackConfigModule extends TopLevel {
  WebpackConfigModule() { getFile().getStem() = "webpack.config" }

  DataFlow::Node getExportedNode() { result = this.(Module).getABulkExportedNode() }

  DataFlow::SourceNode getMainObject() {
    result = getExportedNode().getALocalSource()
    or
    result = getMainObject().(DataFlow::ObjectLiteralNode).getASpreadProperty().getALocalSource()
  }

  DataFlow::SourceNode getResolveObject() {
    result = getMainObject().getAPropertySource("resolve")
    or
    result = getResolveObject().(DataFlow::ObjectLiteralNode).getASpreadProperty().getALocalSource()
  }

  DataFlow::SourceNode getAliasObject() {
    result = getResolveObject().getAPropertySource("alias")
    or
    result = getAliasObject().(DataFlow::ObjectLiteralNode).getASpreadProperty().getALocalSource()
  }

  predicate hasAlias(string name, DataFlow::Node value) {
    getAliasObject().getAPropertyWrite(name).getRhs() = value
  }

  predicate hasAliasValue(string name, string value) {
    exists(DataFlow::Node node |
      hasAlias(name, node) and
      value = [node.getStringValue(), node.asExpr().(PathExpr).getValue()]
    )
  }

  /**
   * Gets the folder affected by this webpack configuration.
   *
   * If there is no sibling `package.json` file, it will use the parent folder
   * if a `package.json` file could be found there.
   */
  Folder getEffectiveFolder() {
    exists(Folder folder | folder = getFile().getParentContainer() |
      if
        not exists(folder.getFile("package.json")) and
        exists(folder.getParentContainer().getFile("package.json"))
      then result = folder.getParentContainer()
      else result = folder
    )
  }
}

private class WebpackPathAlias extends ImportResolution::ScopedPathMapping {
  WebpackConfigModule config;

  WebpackPathAlias() { this = config.getEffectiveFolder() }

  override predicate replaceByPrefix(string oldPrefix, string newPrefix) {
    config.hasAliasValue(oldPrefix, newPrefix)
  }
}
