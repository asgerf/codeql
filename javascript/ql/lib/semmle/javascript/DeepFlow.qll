import javascript
private import semmle.javascript.dataflow.internal.FlowSteps as FlowSteps
private import semmle.javascript.dataflow.internal.StepSummary
private import semmle.javascript.dataflow.TypeTracking
private import internal.CachedStages
private import semmle.javascript.dataflow.internal.DataFlowNode

module Deep {
  private predicate shouldTrack(DataFlow::SourceNode node) {
    not node.getTopLevel().isExterns() and
    not node.getAstNode().isAmbient() and
    not node.getFile().getBaseName().matches("%.d.ts") and
    // only track synthetic return node of async/generator functions, where it represents an allocation
    not exists(Function f |
      DataFlow::functionReturnNode(node, f) and
      not f.isAsyncOrGenerator()
    )
  }

  private predicate initialState(
    DataFlow::SourceNode node, boolean hasCall, boolean hasReturn, boolean promisified,
    int boundArgs
  ) {
    shouldTrack(node) and
    hasReturn = false and
    promisified = false and
    boundArgs = 0 and
    if node instanceof DataFlow::ParameterNode then hasCall = true else hasCall = false
  }

  pragma[nomagic]
  private predicate levelStep(DataFlow::SourceNode pred, DataFlow::SourceNode succ) {
    StepSummary::step(pred, succ, LevelStep())
    or
    any(API::AdditionalUseStep st).step(pred, succ)
    or
    // additional flow `exports` or `module.exports` in `require('m')`
    exists(Import imp | imp.getImportedModuleNode() = succ |
      pred = DataFlow::exportsVarNode(imp.getImportedModule())
      or
      pred = DataFlow::moduleVarNode(imp.getImportedModule()).getAPropertyRead("exports")
    )
  }

  /**
   * Holds if `pred` is cloned or has its properties copied into `succ`.
   */
  pragma[nomagic]
  private predicate massAssignmentStep(DataFlow::SourceNode pred, DataFlow::SourceNode succ) {
    // p -> { ...p }
    exists(ObjectExpr obj |
      succ = obj.flow() and
      pred =
        obj.getAProperty()
            .(SpreadProperty)
            .getInit()
            .(SpreadElement)
            .getOperand()
            .flow()
            .getALocalSource()
    )
  }

  pragma[nomagic]
  private predicate storeStep(DataFlow::SourceNode pred, DataFlow::SourceNode succ, string prop) {
    StepSummary::step(pred, succ, StoreStep(prop))
  }

  pragma[nomagic]
  private predicate loadStep(DataFlow::SourceNode pred, DataFlow::SourceNode succ, string prop) {
    StepSummary::step(pred, succ, LoadStep(prop))
  }

  pragma[nomagic]
  private predicate loadStoreStep(
    DataFlow::SourceNode pred, DataFlow::SourceNode succ, string prop1, string prop2
  ) {
    StepSummary::step(pred, succ, LoadStoreStep(prop1, prop2))
    or
    StepSummary::step(pred, succ, CopyStep(prop1)) and prop2 = prop1
  }

  /**
   * Holds if `pred` can flow to `succ`.
   *
   * `hasCall` and `hasReturn` indicate if the path contains calls and returns, respectively.
   */
  pragma[nomagic]
  DataFlow::SourceNode trackNode(
    DataFlow::SourceNode node, boolean hasCall, boolean hasReturn, boolean promisified,
    int boundArgs
  ) {
    initialState(node, hasCall, hasReturn, promisified, boundArgs) and
    result = node
    or
    exists(DataFlow::SourceNode mid |
      mid = trackNode(node, _, hasReturn, promisified, boundArgs) and
      FlowSteps::callStep(mid.getALocalUse(), result) and
      hasCall = true
    )
    or
    exists(DataFlow::SourceNode mid |
      mid = trackNode(node, false, _, promisified, boundArgs) and
      FlowSteps::returnStep(mid.getALocalUse(), result) and
      hasReturn = true and
      hasCall = false
    )
    or
    levelStep(trackNode(node, hasCall, hasReturn, promisified, boundArgs), result)
    or
    // note: for now, mass assignment steps are simply treated as level steps
    massAssignmentStep(trackNode(node, hasCall, hasReturn, promisified, boundArgs), result)
    or
    exists(boolean call1, boolean call2, boolean return1, boolean return2 |
      derivedPropStep(trackNode(node, call1, return1, promisified, boundArgs), result, call2,
        return2) and
      call1.booleanAnd(return2) = false and
      hasCall = call1.booleanOr(call2) and
      hasReturn = return1.booleanOr(return2)
    )
    or
    // promisification
    exists(Promisify::PromisifyCall promisify |
      trackNode(node, hasCall, hasReturn, false, boundArgs).flowsTo(promisify.getArgument(0)) and
      promisified = true and
      result = promisify
    )
    or
    // partial invocation
    exists(DataFlow::PartialInvokeNode pin, DataFlow::Node pred, int predBoundArgs |
      trackNode(node, hasCall, hasReturn, promisified, predBoundArgs).flowsTo(pred) and
      result = pin.getBoundFunction(pred, boundArgs - predBoundArgs) and
      boundArgs in [0 .. 10]
    )
  }

  /**
   * Holds if `pred` can flow to `succ` via a store/load pair.
   *
   * `hasCall` and `hasReturn` indicate if the path of the carrying object contains
   * calls and returns, respectively.
   */
  pragma[nomagic]
  predicate derivedPropStep(
    DataFlow::SourceNode pred, DataFlow::SourceNode succ, boolean hasCall, boolean hasReturn
  ) {
    exists(DataFlow::SourceNode obj, string prop |
      storeStep(pred, obj, prop) and
      loadStep(trackNode(obj, hasCall, hasReturn, false, 0), succ, prop)
      or
      exists(boolean return1, boolean call1, boolean return2, boolean call2 |
        storeChain(pred, obj, call1, return1, prop) and
        loadStep(trackNode(obj, call2, return2, false, 0), succ, prop) and
        return1.booleanAnd(call2) = false and
        hasCall = call1.booleanOr(call2) and
        hasReturn = return1.booleanOr(return2)
      )
    )
  }

  /**
   * Holds if `pred` can flow to the `prop` property of `succ` via a store followed by one or
   * more load-store steps (not yet finalized by a load step).
   *
   * `hasCall` and `hasReturn` indicate if the paths of the carrying objects contain
   * calls and returns, respectively.
   */
  pragma[nomagic]
  predicate storeChain(
    DataFlow::SourceNode pred, DataFlow::SourceNode succ, boolean hasCall, boolean hasReturn,
    string prop
  ) {
    exists(DataFlow::SourceNode obj, string midProp |
      storeStep(pred, obj, midProp) and
      loadStoreStep(trackNode(obj, hasCall, hasReturn, false, 0), succ, midProp, prop)
      or
      exists(boolean call1, boolean return1, boolean call2, boolean return2 |
        storeChain(pred, obj, call1, return1, midProp) and
        loadStoreStep(trackNode(obj, call2, return2, false, 0), succ, midProp, prop) and
        return1.booleanAnd(call2) = false and
        hasCall = call1.booleanOr(call2) and
        hasReturn = return1.booleanOr(return2)
      )
    )
  }

  cached
  predicate hasFlowTo(DataFlow::SourceNode source, DataFlow::SourceNode dest) {
    trackNode(source, _, _, false, 0) = dest
  }

  class Node extends TNode instanceof DataFlow::SourceNode {
    string toString() { result = this.(DataFlow::SourceNode).toString() }

    predicate hasLocationInfo(
      string filepath, int startline, int startcolumn, int endline, int endcolumn
    ) {
      this.(DataFlow::SourceNode)
          .hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
    }

    cached
    DataFlow::SourceNode getAUse() { hasFlowTo(this, result) }

    cached
    DataFlow::SourceNode getASource() { hasFlowTo(result, this) }

    Node getAPropertyRead(string prop) { result = this.getAUse().getAPropertyRead(prop) }

    Node getAPropertyRead() { result = this.getAUse().getAPropertyRead() }

    DataFlow::PropWrite getAPropertyWrite(string name) {
      result = this.getAUse().getAPropertyWrite(name)
    }

    DataFlow::PropWrite getAPropertyWrite() { result = this.getAUse().getAPropertyWrite() }

    DataFlow::MethodCallNode getAMethodCall(string name) {
      result = this.getAUse().getAMethodCall(name)
    }

    DataFlow::MethodCallNode getAMethodCall() { result = this.getAUse().getAMethodCall() }

    DataFlow::CallNode getACall() { result = this.getAUse().getACall() }
  }
}
