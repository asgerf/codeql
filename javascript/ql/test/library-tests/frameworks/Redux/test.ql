import javascript

query predicate reducerAffectsAccessPath = Redux::isReducerAffectingAccessPath/2;
