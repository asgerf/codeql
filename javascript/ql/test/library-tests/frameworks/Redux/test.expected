| reducers.js:6:10:11:5 | def (member foo (parameter 0 (member combineReducers (member exports (module redux))))) | reducers.js:15:49:15:57 | use (member foo (parameter 0 (parameter 0 (member createSelector (member exports (module reselect)))))) |
| reducers.js:6:10:11:5 | def (member foo (parameter 0 (member combineReducers (member exports (module redux))))) | reducers.js:16:46:16:54 | use (member foo (parameter 0 (parameter 0 (member createSelector (member exports (module reselect)))))) |
